File: ccb-android/app/src/main/java/com/techres/ccb/data/local/entity/BillPrinterConfigEntity.kt
package com.techres.ccb.data.local.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

/**
 * Bill Printer Config Entity - Cấu hình máy in bill
 *
 * Mỗi chi nhánh có thể có nhiều máy in bill (nhiều quầy thu ngân)
 */
@Entity(
    tableName = "bill_printer_configs",
    indices = [
        Index(value = ["branch_id"]),
        Index(value = ["is_active"]),
        Index(value = ["is_default"])
    ]
)
data class BillPrinterConfigEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "branch_id")
    val branchId: String,

    @ColumnInfo(name = "name")
    val name: String, // Tên máy in (VD: "Quầy thu ngân 1")

    @ColumnInfo(name = "description")
    val description: String? = null,

    // ============ CONNECTION CONFIG ============
    @ColumnInfo(name = "connection_type")
    val connectionType: String = "network", // network, bluetooth, usb, sunmi

    @ColumnInfo(name = "printer_ip")
    val printerIp: String? = null,

    @ColumnInfo(name = "printer_port")
    val printerPort: Int = 9100,

    @ColumnInfo(name = "printer_mac")
    val printerMac: String? = null, // Địa chỉ MAC cho Bluetooth

    @ColumnInfo(name = "printer_usb_path")
    val printerUsbPath: String? = null, // USB device path

    // ============ TEMPLATE CONFIG ============
    @ColumnInfo(name = "template_id")
    val templateId: String? = null,

    // ============ PRINT CONFIG ============
    @ColumnInfo(name = "paper_width")
    val paperWidth: Int = 80, // 58 hoặc 80mm

    @ColumnInfo(name = "auto_print_on_payment")
    val autoPrintOnPayment: Boolean = true, // Tự động in khi thanh toán

    @ColumnInfo(name = "print_preview")
    val printPreview: Boolean = false, // Xem trước khi in

    @ColumnInfo(name = "number_of_copies")
    val numberOfCopies: Int = 1,

    @ColumnInfo(name = "cut_paper")
    val cutPaper: Boolean = true,

    @ColumnInfo(name = "open_cash_drawer")
    val openCashDrawer: Boolean = true,

    @ColumnInfo(name = "beep_after_print")
    val beepAfterPrint: Boolean = true,

    // ============ RETRY CONFIG ============
    @ColumnInfo(name = "retry_count")
    val retryCount: Int = 3,

    @ColumnInfo(name = "retry_delay_ms")
    val retryDelayMs: Int = 1000,

    @ColumnInfo(name = "connection_timeout_ms")
    val connectionTimeoutMs: Int = 5000,

    // ============ STATUS ============
    @ColumnInfo(name = "is_default")
    val isDefault: Boolean = false,

    @ColumnInfo(name = "is_active")
    val isActive: Boolean = true,

    @ColumnInfo(name = "sort_order")
    val sortOrder: Int = 0,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    // Sync fields
    @ColumnInfo(name = "sync_status")
    val syncStatus: String = "synced",

    @ColumnInfo(name = "synced_at")
    val syncedAt: String? = null,

    // Local status (not synced)
    @ColumnInfo(name = "is_connected")
    val isConnected: Boolean = false,

    @ColumnInfo(name = "last_print_at")
    val lastPrintAt: String? = null,

    @ColumnInfo(name = "last_error")
    val lastError: String? = null
)

/**
 * Enum for printer connection types
 */
enum class PrinterConnectionType(val value: String) {
    NETWORK("network"),
    BLUETOOTH("bluetooth"),
    USB("usb"),
    SUNMI("sunmi")
}

File: api-dashboard/src/database/entities/bill-printer-config.entity.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  JoinColumn,
  Index,
} from 'typeorm';
import { Branch } from './branch.entity';
import { BillTemplate } from './bill-template.entity';

export enum PrinterConnectionType {
  NETWORK = 'network',
  BLUETOOTH = 'bluetooth',
  USB = 'usb',
  SUNMI = 'sunmi',
}

@Entity('bill_printer_configs')
@Index(['tenantId', 'branchId'])
export class BillPrinterConfig {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ name: 'tenant_id' })
  @Index()
  tenantId: string;

  @Column({ name: 'branch_id' })
  branchId: string;

  @ManyToOne(() => Branch)
  @JoinColumn({ name: 'branch_id' })
  branch: Branch;

  @Column({ length: 100 })
  name: string;

  @Column({ type: 'text', nullable: true })
  description: string;

  // ============ CONNECTION CONFIG ============
  @Column({
    name: 'connection_type',
    type: 'varchar',
    length: 50,
    default: PrinterConnectionType.NETWORK,
  })
  connectionType: string;

  @Column({ name: 'printer_ip', length: 50, nullable: true })
  printerIp: string;

  @Column({ name: 'printer_port', type: 'int', default: 9100 })
  printerPort: number;

  @Column({ name: 'printer_mac', length: 50, nullable: true })
  printerMac: string;

  @Column({ name: 'printer_usb_path', length: 200, nullable: true })
  printerUsbPath: string;

  // ============ TEMPLATE CONFIG ============
  @Column({ name: 'template_id', nullable: true })
  templateId: string;

  @ManyToOne(() => BillTemplate, { nullable: true })
  @JoinColumn({ name: 'template_id' })
  template: BillTemplate;

  // ============ PRINT CONFIG ============
  @Column({ name: 'paper_width', type: 'int', default: 80 })
  paperWidth: number;

  @Column({ name: 'font_size', length: 20, default: 'normal' })
  fontSize: string; // extra_small, small, normal, large, extra_large

  @Column({ name: 'line_spacing', type: 'float', default: 0.7 })
  lineSpacing: number; // 0.3-1.0

  @Column({ name: 'auto_print_on_payment', default: true })
  autoPrintOnPayment: boolean;

  @Column({ name: 'print_preview', default: false })
  printPreview: boolean;

  @Column({ name: 'number_of_copies', type: 'int', default: 1 })
  numberOfCopies: number;

  @Column({ name: 'cut_paper', default: true })
  cutPaper: boolean;

  @Column({ name: 'open_cash_drawer', default: true })
  openCashDrawer: boolean;

  @Column({ name: 'beep_after_print', default: true })
  beepAfterPrint: boolean;

  // ============ RETRY CONFIG ============
  @Column({ name: 'retry_count', type: 'int', default: 3 })
  retryCount: number;

  @Column({ name: 'retry_delay_ms', type: 'int', default: 1000 })
  retryDelayMs: number;

  @Column({ name: 'connection_timeout_ms', type: 'int', default: 5000 })
  connectionTimeoutMs: number;

  // ============ STATUS ============
  @Column({ name: 'is_default', default: false })
  isDefault: boolean;

  @Column({ name: 'is_active', default: true })
  isActive: boolean;

  @Column({ name: 'sort_order', default: 0 })
  sortOrder: number;

  @CreateDateColumn({ name: 'created_at' })
  createdAt: Date;

  @UpdateDateColumn({ name: 'updated_at' })
  updatedAt: Date;
}
